File: kenkenpa-compiler/src/main/java/com/github/shiraji/kenkenpa/compiler/model/HopModel.java
package com.github.shiraji.kenkenpa.compiler.model;

import com.github.shiraji.kenkenpa.annotations.Hop;
import com.github.shiraji.kenkenpa.annotations.Hops;
import com.github.shiraji.kenkenpa.compiler.helper.KenKenPaHelper;

import javax.lang.model.element.Element;
import java.lang.annotation.Annotation;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class HopModel {
    Map<Element, Annotation> mHopMap = new HashMap<>();

    public boolean isValidHops(Element element) {
        Hops hops = element.getAnnotation(Hops.class);
        if(hops == null) {
            return true;
        }

        List<String> hopsFrom = new ArrayList<>();
        for (Hop hop : hops.value()) {
            if (hopsFrom.contains(hop.from())) {
                return false;
            }
        }
        return true;
    }

    private void handleHopAnnotation(Element element) {
        Hop hop = element.getAnnotation(Hop.class);
        if (hop == null) {
            return;
        }
        mHopMap.put(element, hop);
    }

    private void handleHopsMethods(Element element) {
        Hops hops = element.getAnnotation(Hops.class);
        if (hops == null) {
            return;
        }
        mHopMap.put(element, hops);
    }
}

File: kenkenpa-compiler/src/main/kotlin/com/github/shiraji/kenkenpa/compiler/model/HopModel.kt
package com.github.shiraji.kenkenpa.compiler.model

import com.github.shiraji.kenkenpa.annotations.Hop
import com.github.shiraji.kenkenpa.annotations.Hops
import com.github.shiraji.kenkenpa.compiler.helper.KenKenPaHelper

import javax.lang.model.element.Element
import java.util.ArrayList
import java.util.HashMap

class HopModel {
    internal var mHopMap: MutableMap<Element, Annotation> = HashMap()

    fun isValidHops(element: Element): Boolean {
        val hops = element.getAnnotation(Hops::class.java) ?: return true
        val hopsFrom = ArrayList<String>()
        hops.value.forEach {
            if (hopsFrom.contains(it.from)) {
                return false
            }
            hopsFrom.add(it.from)
        }
        return true
    }

    private fun handleHopAnnotation(element: Element) {
        val hop = element.getAnnotation(Hop::class.java) ?: return
        mHopMap.put(element, hop)
    }

    private fun handleHopsMethods(element: Element) {
        val hops = element.getAnnotation(Hops::class.java) ?: return
        mHopMap.put(element, hops)
    }
}
